src/jrd/recsrc/TableValueFunctionScan.cpp

```diff
@@ -0,0 +1,277 @@
+
+	if (!level)
+		plan += ")";
+}
+
+void TableValueFunctionScan::assignParameter(thread_db* tdbb, dsc* fromDesc, const dsc* toDesc,
+											 SSHORT toId, Record* record) const
+{
+	dsc toDescValue = *toDesc;
+	toDescValue.dsc_address = record->getData() + (IPTR)toDesc->dsc_address;
+
+	if (fromDesc->isNull())
+	{
+		record->setNull(toId);
+		return;
+	}
+	else
+		record->clearNull(toId);
+
+	if (!DSC_EQUIV(fromDesc, &toDescValue, false))
+	{
+		MOV_move(tdbb, fromDesc, &toDescValue);
+		return;
+	}
+
+	memcpy(toDescValue.dsc_address, fromDesc->dsc_address, fromDesc->dsc_length);
+}
+
+UnlistFunctionScan::UnlistFunctionScan(CompilerScratch* csb, StreamType stream, const string& alias,
+									   ValueListNode* list)
+	: TableValueFunctionScan(csb, stream, alias), m_inputList(list)
+{
+}
+
+void UnlistFunctionScan::internalOpen(thread_db* tdbb) const
+{
+	const auto request = tdbb->getRequest();
+	const auto impure = request->getImpure<Impure>(m_impure);
+
+	impure->irsb_flags = irsb_open;
+
+	const auto rpb = &request->req_rpb[m_stream];
+
+	MemoryPool& pool = *tdbb->getDefaultPool();
+	impure->m_recordBuffer = FB_NEW_POOL(pool) RecordBuffer(pool, m_format);
+
+	Record* const record = VIO_record(tdbb, rpb, m_format, &pool);
+
+	rpb->rpb_number.setValue(BOF_NUMBER);
+
+	fb_assert(m_inputList->items.getCount() >= UNLIST_INDEX_LAST);
+
+	auto valueItem = m_inputList->items[UNLIST_INDEX_STRING];
+	const auto valueDesc = EVL_expr(tdbb, request, valueItem);
+	if (valueDesc == nullptr)
+	{
+		rpb->rpb_number.setValid(true);
+		return;
+	}
+
+	auto separatorItem = m_inputList->items[UNLIST_INDEX_SEPARATOR];
+	const auto separatorDesc = EVL_expr(tdbb, request, separatorItem);
+	if (separatorDesc == nullptr)
+	{
+		rpb->rpb_number.setValid(false);
+		return;
+	}
+
+	auto toDesc = m_format->fmt_desc.begin();
+	fb_assert(toDesc);
+
+	const auto textType = toDesc->getTextType();
+
+	auto setStringToRecord = [&] (string str, USHORT length = 0)
+	{
+		if (length == 0)
+			length = str.length();
+		dsc fromDesc;
+		fromDesc.makeText(length, textType, (UCHAR*)str.c_str());
+		assignParameter(tdbb, &fromDesc, toDesc, 0, record);
+		impure->m_recordBuffer->store(record);
+	};
+
+	string separatorStr(MOV_make_string2(tdbb, separatorDesc, textType, true));
+
+	if (separatorStr.isEmpty())
+	{
+		string valueStr(MOV_make_string2(tdbb, valueDesc, textType, false));
+		setStringToRecord(valueStr);
+		return;
+	}
+
+	if (valueDesc->isBlob())
+	{
+		blb* blob = blb::open(tdbb, request->req_transaction,
+						reinterpret_cast<bid*>(valueDesc->dsc_address));
+
+		string resultStr;
+
+		while (!(blob->blb_flags & BLB_eof))
+		{
+			UCHAR buffer[BUFFER_LARGE];
+			const auto length = blob->BLB_get_data(tdbb, buffer, sizeof(buffer), false);
+			if (length == 0)
+				continue;
+
+			string valueStr(buffer, length);
+
+			auto end = AbstractString::npos;
+			do
+			{
+				auto size = end = valueStr.find(separatorStr);
```

Comment:
Is this code compatible with multi-character separator and the list string split in different blob segments?

Reply:
No, it's not. To be improved, I guess. While `LIST` would never split the separator, the input may be provided from the different source as well and we cannot predict the layout.

Reply:
@asfernandes, do we generally consider OK to read/process the blob as a whole? If so, this would be the easiest solution (and IIRC it worked this way in our original implementation) and we do the same in some built-in functions. Or should it be considered a bad practice to be revisited/fixed and thus the new code should be able to process blobs in chunks?

Reply:
It requires to move the processing from `open()` to `getRecord()`. Currently implemented differently, but it could be worth the effort.

Reply:
I wrote my UDR Split exactly like this. Not perfect, but works well https://github.com/sim1984/split_udr/blob/master/src/SplitProcedure.cpp

Reply:
> @asfernandes, do we generally consider OK to read/process the blob as a whole? If so, this would be the easiest solution (and IIRC it worked this way in our original implementation) and we do the same in some built-in functions. Or should it be considered a bad practice to be revisited/fixed and thus the new code should be able to process blobs in chunks?

For this function, the cost/benefit to incrementally read the input is very clear for me.

Reply:
Shouldn't it depend on `OPTIMIZE FOR {FIRST | ALL} ROWS`? I.e. process/return records in chunks for `ALL ROWS` (not necessarily read/cache the whole blob, maybe just some sliding buffer) but one by one for `FIRST ROWS`?

Reply:
> Shouldn't it depend on `OPTIMIZE FOR {FIRST | ALL} ROWS`? I.e. process/return records in chunks for `ALL ROWS` (not necessarily read/cache the whole blob, maybe just some sliding buffer) but process/return records one by one for `FIRST ROWS`?

I do not see a clear relation of `OPTIMIZE FOR` and `UNNEST`. `OPTIMIZE FOR` is top level query feature, while `UNNEST` may be in inner queries.

Reply:
I mean the mechanics, not the clause itself. For example, a subquery with a `FIRST` clause implies `FOR FIRST ROWS` when compiling/optimizing its own RSE.

Reply:
The current implementation has two drawbacks:
1. Increased memory consumption due to result buffering
2. Slowing down queries with the FIRST ROWS strategy

In general, it is not very clear why TableValueFunction is a buffered data source. It seems to be related to primary, and all our primary sources are cover. If you need to buffer the result, then the optimizer of the higher level does it.

Reply:
These issues are being addressed.